
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="../../assets/images/favicon.ico">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-8.3.9">
    
    
      
        <title>Example-driven guide - LoVis4u</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.1dff34a1.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.cbb835fc.min.css">
        
      
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_markdown_exec_pyodide.css">
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../css/material.css">
    
      <link rel="stylesheet" href="../../css/mkdocstrings.css">
    
      <link rel="stylesheet" href="https://unpkg.com/katex@0/dist/katex.min.css">
    
    <script>__md_scope=new URL("../..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  


  <script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-4JWMGFSKXS"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){this.value&&gtag("event","search",{search_term:this.value})}),"undefined"!=typeof location$&&location$.subscribe(function(e){gtag("config","G-4JWMGFSKXS",{page_path:e.pathname})})})</script>
  <script async src="https://www.googletagmanager.com/gtag/js?id=G-4JWMGFSKXS"></script>


    
    
   <link href="../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
    html.glightbox-open { overflow: initial; height: 100%; }
    .gslide-title { margin-top: 0px; user-select: text; }
    .gslide-desc { color: #666; user-select: text; }
    .gslide-image img { background: white; }
    .gscrollbar-fixer { padding-right: 15px; }
    .gdesc-inner { font-size: 0.75rem; }
    body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
    body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
    body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}</style> <script src="../../assets/javascripts/glightbox.min.js"></script></head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="lovis4u_brown" data-md-color-accent="lovis4u_yellow">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#example-driven-guide" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="LoVis4u" class="md-header__button md-logo" aria-label="LoVis4u" data-md-component="logo">
      
  
  <?xml version="1.0" encoding="UTF-8"?>
<svg id="Layer_1" data-name="Layer 1" xmlns="http://www.w3.org/2000/svg" width="500" height="500" viewBox="0 0 500 500">
  <path d="M233.703576745615464,427.256416679763788c7.333369773565209,4.493880934460321,16.504100586736058,4.58153700476214,23.921944907303441.228507746331161,50.132626006159626-29.419434862202252,238.137964177803951-148.749554434098172,200.44862634459605-263.45746735486955-.038701212821252-.160953878998953-.09111622448836-.295231716696435-.143214562194771-.456303087769811-13.348463600363175-43.859509781060297-53.757954843338666-75.923165964325563-101.793264469624773-76.344433085516357-27.075997864574674-.23745506717114-51.89797185122552,9.610385093721561-70.942033621216979,26.067702736280808l-.01351461696504.013279632835292c-.013513900568796.013197943029809-.040343952648982.026360800184193-.053858569615841.039640433006753-1.48227180887443,1.124700009913795-25.064705234497524,19.201953708738074-37.35273767684339,39.104905121658703-10.847787872680783-19.168879117740289-29.646930734423222-35.221834687822593-35.175457092377655-39.740973548275178-18.842202906597777-17.244625280741275-43.839161852159123-27.127838666213393-71.303267889790732-27.36869742151066-48.048788440949465-.421385329678742-89.005043205677794,31.397142829636323-103.120685642969875,75.01579385540299-.054996971637593.1601321508806-.114976038024906-.630068064120678-.156494195023697-.46981770473667C-1.665428560296277,273.921462364258332,184.128852320982332,396.877080057964122,233.703576745615464,427.256416679763788Z" style="fill: none; stroke: #fff; stroke-miterlimit: 10; stroke-width: 43px;"/>
</svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            LoVis4u
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Example-driven guide
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="lovis4u_brown" data-md-color-accent="lovis4u_yellow"  aria-hidden="true"  type="radio" name="__palette" id="__palette_1">
          
        
      </form>
    
    
    
    
      <div class="md-header__source">
        <a href="https://github.com/art-egorov/lovis4u.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-tabs__inner md-grid">
    <ul class="md-tabs__list">
      
        
  
  


  <li class="md-tabs__item">
    <a href="../.." class="md-tabs__link">
      Home
    </a>
  </li>

      
        
  
  


  <li class="md-tabs__item">
    <a href="../../Pipeline/pipeline/" class="md-tabs__link">
      Pipeline description
    </a>
  </li>

      
        
  
  
    
  


  <li class="md-tabs__item">
    <a href="./" class="md-tabs__link md-tabs__link--active">
      Example-driven guide
    </a>
  </li>

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../Parameters/cmd_parameters/" class="md-tabs__link">
        Parameters
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../API/usage_examples/" class="md-tabs__link">
        API
      </a>
    </li>
  

      
        
  
  


  <li class="md-tabs__item">
    <a href="../../VersionLog/versions/" class="md-tabs__link">
      Version log
    </a>
  </li>

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="LoVis4u" class="md-nav__button md-logo" aria-label="LoVis4u" data-md-component="logo">
      
  
  <?xml version="1.0" encoding="UTF-8"?>
<svg id="Layer_1" data-name="Layer 1" xmlns="http://www.w3.org/2000/svg" width="500" height="500" viewBox="0 0 500 500">
  <path d="M233.703576745615464,427.256416679763788c7.333369773565209,4.493880934460321,16.504100586736058,4.58153700476214,23.921944907303441.228507746331161,50.132626006159626-29.419434862202252,238.137964177803951-148.749554434098172,200.44862634459605-263.45746735486955-.038701212821252-.160953878998953-.09111622448836-.295231716696435-.143214562194771-.456303087769811-13.348463600363175-43.859509781060297-53.757954843338666-75.923165964325563-101.793264469624773-76.344433085516357-27.075997864574674-.23745506717114-51.89797185122552,9.610385093721561-70.942033621216979,26.067702736280808l-.01351461696504.013279632835292c-.013513900568796.013197943029809-.040343952648982.026360800184193-.053858569615841.039640433006753-1.48227180887443,1.124700009913795-25.064705234497524,19.201953708738074-37.35273767684339,39.104905121658703-10.847787872680783-19.168879117740289-29.646930734423222-35.221834687822593-35.175457092377655-39.740973548275178-18.842202906597777-17.244625280741275-43.839161852159123-27.127838666213393-71.303267889790732-27.36869742151066-48.048788440949465-.421385329678742-89.005043205677794,31.397142829636323-103.120685642969875,75.01579385540299-.054996971637593.1601321508806-.114976038024906-.630068064120678-.156494195023697-.46981770473667C-1.665428560296277,273.921462364258332,184.128852320982332,396.877080057964122,233.703576745615464,427.256416679763788Z" style="fill: none; stroke: #fff; stroke-miterlimit: 10; stroke-width: 43px;"/>
</svg>

    </a>
    LoVis4u
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/art-egorov/lovis4u.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../Pipeline/pipeline/" class="md-nav__link">
        Pipeline description
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Example-driven guide
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Example-driven guide
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#example-run-with-default-parameters" class="md-nav__link">
    Example run with default parameters
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#using-the-locus-annotation-table" class="md-nav__link">
    Using the locus annotation table
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#using-the-feature-annotation-table" class="md-nav__link">
    Using the feature annotation table
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#other-features" class="md-nav__link">
    Other features
  </a>
  
    <nav class="md-nav" aria-label="Other features">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#category-colour-and-annotation" class="md-nav__link">
    Category colour and annotation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-a-scale-line-instead-of-individual-x-axis" class="md-nav__link">
    Using a scale line instead of individual x-axis
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#highlighting-conserved-genes-instead-of-variable" class="md-nav__link">
    Highlighting conserved genes instead of variable
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#specifying-figure-width" class="md-nav__link">
    Specifying figure width
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4">
          Parameters
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Parameters" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Parameters
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../Parameters/cmd_parameters/" class="md-nav__link">
        Command-line parameters
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../Parameters/config_parameters/" class="md-nav__link">
        Configuration file parameters
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5">
          API
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="API" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          API
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../API/usage_examples/" class="md-nav__link">
        Usage examples
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../API/package/" class="md-nav__link">
        Library
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../VersionLog/versions/" class="md-nav__link">
        Version log
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                


<h1 id="example-driven-guide">Example-driven guide</h1>
<p>Here we show usage examples of the lovis4u command-line interface. Through this guide we will show you step-by-step how to optimise your visualisation starting from default parameters.</p>
<p><strong>Before start:</strong> The necessary sample data as well as adjustable tool configuration files are provided by lovis4u at the post-install step with the command:  <br />
<code>lovis4u --data</code> This copies the <em>lovis4u_data</em> folder to your working directory.</p>
<p><strong>If you work on a Linux machine:</strong> after installation you should run: <code>lovis4u --linux</code><br />
This command replaces the tool path (for MMseqs2) in the pre-made config files from the MacOS version (default) to the Linux version.
If you run this command for fun and want to change it back you can use <code>lovis4u --mac</code>.</p>
<p><ins>For demonstration we will use pharokka generated gff files with the sequences of five Enterobacteria phages.
Gff files are stored at: lovis4u_data/guide/gff_files.</ins>    <br />
The main difference between pharokka generated gff files and regular gff3 (for ex. which you can download from the NCBI) is that in addition to the annotation rows the annotation contains the corresponding nucleotide sequence in fasta format.</p>
<hr />
<h2 id="example-run-with-default-parameters">Example run with default parameters</h2>
<p>Let's start with running lovis4u without using any optional arguments. The only mandatory argument is a folder path containing pharokka generated gff (<code>-gff</code>) files or genbank files (<code>-gb</code>).   </p>
<div class="language-sh highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>lovis4u<span class="w"> </span>-gff<span class="w"> </span>lovis4u_data/guide/gff_files
</span></code></pre></div>
<p>Running this command will create an output folder named <em>lovis4u_{current_date}</em> (e.g. <em>lovis4u_2024_04_28-16_36</em>) containing the results files. The name of the output folder can be changed with <code>-o &lt;output_folder_name&gt;</code>.</p>
<details class="card_hearts">
<summary><strong>Output folder structure</strong></summary>
<ul>
<li><em>lovis4u.pdf</em> - vector graphic output (file name can be changed with <code>--pdf-name &lt;filename&gt;</code> parameter)  </li>
<li><em>locus_annotation_table.tsv</em> - table containing annotation (sequence_id, length, coordinates, etc.) for each locus.  </li>
<li><em>feature_annotation_table.tsv</em> - table containing annotation (feature_id, locus_id, coordinates, etc.) for each feature (e.g. CDS)  </li>
<li><em>mmseqs</em> (folder)  <ul>
<li><em>DB</em> - folder with MMseqs databases.  </li>
<li><em>mmmseqs_clustering.tsv</em> - table with proteomes clustering results.  </li>
<li><em>mmmseqs_(stdout/stderr).txt</em> - mmseqs logs.  </li>
<li><em>input_proteins.fa</em> - fasta file with all annotated protein sequences (input to mmseqs).  </li>
</ul>
</li>
<li><em>proteome_similarity_matrix.tsv</em> - pairwise proteome similarity scores indicating the fraction of shared proteins homologues.  </li>
</ul>
</details>
<p><a class="glightbox" href="img/lovis4u_default_1.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="f1" loading="lazy" src="img/lovis4u_default_1.png" width="100%" /></a>  </p>
<p><strong>By default, lovis4u uses the following data pre-processing steps:</strong></p>
<ol>
<li>The full length of each input sequence is taken for analysis (this can be adjusted using the locus annotation table, see below).. </li>
<li>All proteins annotated to be encoded on input sequences are used as input for MMseqs2 clustering (can be deactivated with --mmseqs-off parameter). MMseqs2 arguments can be adjusted using config file. Proteins clustered together are considered to be a set of homologues. Based these results, the "group" attribute of each CDS is set.</li>
<li>Taking into account the information about the set of homologues from the previous step, lovis4u applies similarity based hierarchical clustering of proteomes, which it uses to find the optimal order for visualisation and sets the "group" attribute for each locus. The purpose is to group together only related proteomes (keeping the average proteome set similarity &gt; ~80%). This step can be skipped with the <code>-cl-off, --clust_loci-off</code> parameter.</li>
<li>Defining the feature attribute "group_type". This allows the use of parameters that refer to a particular set of feature groups (e.g. set colour or show labels only for "group_type" = "variable"). By default it sets group_type "variable" for CDS features that are found in less than 0.25 of loci within the locus group, and "conserved" for others that are less variable. </li>
<li>Setting feature colour based on the feature "group" attribute (can be deactivated with -sgc-off or --set-group-colour-off). By default loVis4u sets distinct colours only for features with group_type "variable". You can change this with -sgcf or --set-group-colour-for. For example, if you want to set colour only for features with group_type "conserved", run <code>--set-group-colour-for conserved</code>. </li>
<li>Defining labels to be shown. By default, lovis4u shows all labels for "variable" features and only the first occurrence for "conserved" features. You can show all labels with <code>--show-all-feature-labels</code> or specify group types for which all labels will be shown with <code>-sflf, --show-feature-label-for &lt;classes&gt;</code>. Additionally, by default lovis4u ignores the following labels: hypothetical protein, unknown protein. The list of ignored labels can be set with <code>-ifl, --ignored-feature-labels &lt;feature_label1 [feature_label2 ...]&gt;</code>. The list with this argument can be left empty to not filter out labels by their name.</li>
</ol>
<p><strong>The following parameters are particularly useful for basic runs:</strong></p>
<ul>
<li><code>--reorient_loci</code> - Auto re-orient loci (set new strands) if they are not matched.  This function tries to maximise co-orientation of homologous features.</li>
<li><code>-hl</code>, <code>--homology-links</code> - Draw homology link track (Sankey graph).</li>
<li><code>-o &lt;name&gt;</code> - Output dir name.  </li>
</ul>
<p>While loci in our test set are already correctly orientated, let's add -hl parameter to update the output.</p>
<p><div class="language-sh highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>lovis4u<span class="w"> </span>-gff<span class="w"> </span>lovis4u_data/guide/gff_files<span class="w">  </span>-hl<span class="w"> </span>-o<span class="w"> </span>lovis4u_output
</span></code></pre></div>
<a class="glightbox" href="img/lovis4u_default_hl.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="f2" loading="lazy" src="img/lovis4u_default_hl.png" width="100%" /></a>  </p>
<h2 id="using-the-locus-annotation-table">Using the locus annotation table</h2>
<p>As mentioned above, the full length of each locus is taken for visualisation by default. However, you can also specify the sub-coordinates of multiple regions for each locus to be shown. The coordinates together with other information about each locus can be specified in the locus annotation table and used as input with  <code>-laf</code> or <code>--locus-annotation-file</code> parameter. </p>
<p>Additionally, after each run lovis4u saves the <em>locus_annotation_table.tsv</em> with annotation parameters used in this particular run. If no table was specified by input then all annotation columns are set with default values.  </p>
<p><em>Default table generated from previous runs:</em></p>
<table>
<thead>
<tr>
<th style="text-align: left;">sequence_id</th>
<th style="text-align: right;">length</th>
<th style="text-align: left;">coordinates</th>
<th style="text-align: right;">circular</th>
<th style="text-align: left;">description</th>
<th style="text-align: right;">order</th>
<th style="text-align: right;">group</th>
<th style="text-align: right;">initial_order</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">RefSeq_NC_005056.1</td>
<td style="text-align: right;">32684</td>
<td style="text-align: left;">1:32684:1</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">Escherichia phage Wphi</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">3</td>
</tr>
<tr>
<td style="text-align: left;">RefSeq_NC_049461.1</td>
<td style="text-align: right;">31834</td>
<td style="text-align: left;">1:31834:1</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">Bacteriophage R18C</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
<tr>
<td style="text-align: left;">RefSeq_NC_001895.1</td>
<td style="text-align: right;">33593</td>
<td style="text-align: left;">1:33593:1</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">Enterobacteria phage P2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
</tr>
<tr>
<td style="text-align: left;">RefSeq_NC_049457.1</td>
<td style="text-align: right;">33807</td>
<td style="text-align: left;">1:33807:1</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">Escherichia phage vB_EcoM-12474III</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">4</td>
</tr>
<tr>
<td style="text-align: left;">RefSeq_NC_028943.1</td>
<td style="text-align: right;">29237</td>
<td style="text-align: left;">1:29237:1</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">Escherichia phage pro483</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
</tr>
</tbody>
</table>
<p>The locus annotation table can also be used as input for customising the regions and orientations of the sequences shown. After the default run we can take the output locus_annotation_table and edit the information we want to change.<br />
<strong>Important to note</strong>: it is possible to input a table containing only a subset of columns (only the sequence_id column is essential). For other columns or empty cells, lovis4u will set the default values.</p>
<p>For example, let's use this table as input:</p>
<table>
<thead>
<tr>
<th style="text-align: left;">sequence_id</th>
<th style="text-align: left;">coordinates</th>
<th style="text-align: right;">order</th>
<th style="text-align: right;">group</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">RefSeq_NC_005056.1</td>
<td style="text-align: left;">18320:24320:1,31684:32684:1,1:9360:1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
</tr>
<tr>
<td style="text-align: left;">RefSeq_NC_049461.1</td>
<td style="text-align: left;">16745:22745:1,30834:31834:1,1:7016:1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
<tr>
<td style="text-align: left;">RefSeq_NC_001895.1</td>
<td style="text-align: left;">18294:24294:1,32593:33593:1,1:8460:1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
</tr>
<tr>
<td style="text-align: left;">RefSeq_NC_049457.1</td>
<td style="text-align: left;">19308:25308:1,32807:33807:1,1:9474:1</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">1</td>
</tr>
<tr>
<td style="text-align: left;">RefSeq_NC_028943.1</td>
<td style="text-align: left;">14839:20839:1,28237:29237:1,1:6321:1</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">1</td>
</tr>
</tbody>
</table>
<p>Here we specified only the coordinates, order and group for each locus. Order and group are also specified and kept since it is logical to use clustering results defined during the full locus length run, and to turn off a new attempt to cluster sequences. This is done with the <code>-cl-off, --clust_loci-off</code> parameter.   </p>
<p><strong>Format for coordinates specification:</strong> comma-separated list of start:end:strand. Start and end are in 1-based format, strand: 1 for plus strand and -1 for minus.</p>
<p>The table can be found in the guide folder: <em>lovis4u_data/guide/locus_annotation_table_alt.tsv</em></p>
<p>Now we can run: 
<div class="language-sh highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>lovis4u<span class="w"> </span>-gff<span class="w"> </span>lovis4u_data/guide/gff_files<span class="w">  </span>-hl<span class="w"> </span>-o<span class="w"> </span>lovis4u_output<span class="w"> </span><span class="se">\</span>
</span><span id="__span-2-2"><a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a><span class="w">    </span>--locus-annotation-file<span class="w"> </span>lovis4u_data/guide/locus_annotation_table_alt.tsv<span class="w"> </span>-cl-off
</span></code></pre></div>
<a class="glightbox" href="img/lovis4u_window_1.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="f3" loading="lazy" src="img/lovis4u_window_1.png" width="100%" /></a>  </p>
<h2 id="using-the-feature-annotation-table">Using the feature annotation table</h2>
<p>Similarly to the locus annotation table, we can use the the feature annotation table to adjust the feature visualisation parameters. Here, feature usually refers to an open reading frame, each of which has its own ID.  After each run, a feature_annotation_table.tsv file is saved in your output folder. Below you can see header of default table created with default parameter run.</p>
<table>
<thead>
<tr>
<th style="text-align: left;">feature_id</th>
<th style="text-align: left;">locus_id</th>
<th style="text-align: left;">coordinates</th>
<th style="text-align: left;">feature_type</th>
<th style="text-align: left;">name</th>
<th style="text-align: left;">group</th>
<th style="text-align: left;">group_type</th>
<th style="text-align: left;">category</th>
<th style="text-align: left;">fill_colour</th>
<th style="text-align: left;">stroke_colour</th>
<th style="text-align: right;">show_label</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">RefSeq_NC_001895.1_CDS_0001</td>
<td style="text-align: left;">RefSeq_NC_001895.1</td>
<td style="text-align: left;">64:2286:1</td>
<td style="text-align: left;">CDS</td>
<td style="text-align: left;">nicking at origin of replication</td>
<td style="text-align: left;">RefSeq_NC_005056.1_CDS_0001</td>
<td style="text-align: left;">conserved</td>
<td style="text-align: left;">"DNA, RNA and nucleotide metabolism"</td>
<td style="text-align: left;">default</td>
<td style="text-align: left;">default</td>
<td style="text-align: right;">0</td>
</tr>
<tr>
<td style="text-align: left;">RefSeq_NC_001895.1_CDS_0002</td>
<td style="text-align: left;">RefSeq_NC_001895.1</td>
<td style="text-align: left;">2283:2612:1</td>
<td style="text-align: left;">CDS</td>
<td style="text-align: left;">hypothetical protein</td>
<td style="text-align: left;">RefSeq_NC_005056.1_CDS_0002</td>
<td style="text-align: left;">conserved</td>
<td style="text-align: left;">unknown function</td>
<td style="text-align: left;">default</td>
<td style="text-align: left;">default</td>
<td style="text-align: right;">0</td>
</tr>
<tr>
<td style="text-align: left;">RefSeq_NC_001895.1_CDS_0003</td>
<td style="text-align: left;">RefSeq_NC_001895.1</td>
<td style="text-align: left;">2651:3412:-1</td>
<td style="text-align: left;">CDS</td>
<td style="text-align: left;">hypothetical protein</td>
<td style="text-align: left;">RefSeq_NC_001895.1_CDS_0003</td>
<td style="text-align: left;">variable</td>
<td style="text-align: left;">unknown function</td>
<td style="text-align: left;">#aa9b35</td>
<td style="text-align: left;">default</td>
<td style="text-align: right;">0</td>
</tr>
<tr>
<td style="text-align: left;">RefSeq_NC_001895.1_CDS_0004</td>
<td style="text-align: left;">RefSeq_NC_001895.1</td>
<td style="text-align: left;">3587:5347:-1</td>
<td style="text-align: left;">CDS</td>
<td style="text-align: left;">hypothetical protein</td>
<td style="text-align: left;">RefSeq_NC_001895.1_CDS_0004</td>
<td style="text-align: left;">variable</td>
<td style="text-align: left;">unknown function</td>
<td style="text-align: left;">#f4767a</td>
<td style="text-align: left;">default</td>
<td style="text-align: right;">0</td>
</tr>
<tr>
<td style="text-align: left;">RefSeq_NC_001895.1_CDS_0005</td>
<td style="text-align: left;">RefSeq_NC_001895.1</td>
<td style="text-align: left;">5347:5517:-1</td>
<td style="text-align: left;">CDS</td>
<td style="text-align: left;">hypothetical protein</td>
<td style="text-align: left;">RefSeq_NC_001895.1_CDS_0005</td>
<td style="text-align: left;">variable</td>
<td style="text-align: left;">unknown function</td>
<td style="text-align: left;">#f47763</td>
<td style="text-align: left;">default</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
<p>As with the locus annotation table, we can tweak the parameters and use this as input in subsequent runs.  This is done with the parameter <code>-faf, --feature-annotation-file</code>, a table that contains only the subset of annotation columns that are of interest. For example, you can specify a new label and colour for those particular coding sequences (CDS) that you are interested in.</p>
<table>
<thead>
<tr>
<th style="text-align: left;">feature_id</th>
<th style="text-align: left;">name</th>
<th style="text-align: left;">fill_colour</th>
<th style="text-align: right;">show_label</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">RefSeq_NC_001895.1_CDS_0001</td>
<td style="text-align: left;">Updated Name :)</td>
<td style="text-align: left;">#e54e86</td>
<td style="text-align: right;">1</td>
</tr>
</tbody>
</table>
<p>This table can be found in the guide folder: <em>lovis4u_data/guide/feature_annotation_table_alt.tsv</em></p>
<p><div class="language-sh highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>lovis4u<span class="w"> </span>-gff<span class="w"> </span>lovis4u_data/guide/gff_files<span class="w">  </span>-hl<span class="w"> </span>-o<span class="w"> </span>lovis4u_output<span class="w"> </span><span class="se">\</span>
</span><span id="__span-3-2"><a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a><span class="w">    </span>--locus-annotation-file<span class="w"> </span>lovis4u_data/guide/locus_annotation_table_alt.tsv<span class="w"> </span>-cl-off<span class="w"> </span><span class="se">\</span>
</span><span id="__span-3-3"><a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a><span class="w">    </span>--feature-annotation-file<span class="w"> </span>lovis4u_data/guide/feature_annotation_table_alt.tsv
</span></code></pre></div>
<a class="glightbox" href="img/lovis4u_updated_name.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="f4" loading="lazy" src="img/lovis4u_updated_name.png" width="100%" /></a>  </p>
<p>Here it is also <strong>important to note</strong> that if you don't use the feature attribute "group_type" defined by the full locus run, then this will be re-defined by mmseqs using on the subset of proteins found only in the specified coordinates given in the locus annotation table. You can deactivate re-definition of group types with <code>--find-variable-off</code> and <code>--mmseqs-off</code>). This avoids a situation where an ORF is labelled "variable", when in reality it is encoded in each proteome, but outside of the coordinates shown.</p>
<h2 id="other-features">Other features</h2>
<h3 id="category-colour-and-annotation">Category colour and annotation</h3>
<p>Using the parameter <code>--set-category-colour</code> you can use the functional category annotation column for features. The functional category is shown by coloured lines below each ORF. By default it was designed to parse the PHROGs category annotation for proteins and retrieve information about the category in the "function" qualifiers in Genbank or GFF files (qualifiers used can be changed in config file). However, you can set the category for each CDS using the above feature annotation table and the "category" column. Additionally, you can set the colour codes for your categories using <code>--category-colour-table</code>. For categories that are not found in a table, a random colour will be set. By default, lovis4u uses a pre-made colour table found in <em>lovis4u_data/category_colours.tsv</em>.</p>
<p><div class="language-sh highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>lovis4u<span class="w"> </span>-gff<span class="w"> </span>lovis4u_data/guide/gff_files<span class="w"> </span>-hl<span class="w"> </span>--set-category-colour
</span></code></pre></div>
<a class="glightbox" href="img/lovis4u_category_colour.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="f5" loading="lazy" src="img/lovis4u_category_colour.png" width="100%" /></a>  </p>
<h3 id="using-a-scale-line-instead-of-individual-x-axis">Using a scale line instead of individual x-axis</h3>
<p>Using the parameter <code>--hide-x-axis</code> you can deactivate visualisation of the x-axis for each locus track and instead, if you wish, draw a scale line track below. This is done using with <code>-slt, --scale-line-track</code>.</p>
<p><div class="language-sh highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>lovis4u<span class="w"> </span>-gff<span class="w"> </span>lovis4u_data/guide/gff_files<span class="w"> </span>-hl<span class="w"> </span>--hide-x-axis<span class="w"> </span>--scale-line-track
</span></code></pre></div>
<a class="glightbox" href="img/lovis4u_scale.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="f6" loading="lazy" src="img/lovis4u_scale.png" width="100%" /></a>  </p>
<h3 id="highlighting-conserved-genes-instead-of-variable">Highlighting conserved genes instead of variable</h3>
<p>For many analysis purposes (e.g. conserved neighbourhood visualisation), it is more useful to colourise conserved gene clusters instead of the variable genes. This can be easily switched in lovis4u using the <code>--set-group-colour-for</code> parameter. Te default is to colour by variable but using <code>--set-group-colour-for conserved</code> will change it to the opposite mode.<br />
<strong>Note</strong> that if you have other feature group set in your feature annotation table and you want to set auto-colourising for them as well, you can specify them in a space separated list with this argument (e.g. <code>--set-group-colour-for conserved your_group_1 your_group_2</code>).</p>
<div class="language-sh highlight"><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>lovis4u<span class="w"> </span>-gff<span class="w"> </span>lovis4u_data/guide/gff_files<span class="w"> </span>-hl<span class="w"> </span>--set-group-colour-for<span class="w"> </span>conserved
</span></code></pre></div>
<p><a class="glightbox" href="img/lovis4u_conserved_colourised.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="f7" loading="lazy" src="img/lovis4u_conserved_colourised.png" width="100%" /></a>  </p>
<p><strong>Note:</strong>By default, the colours for groups are randomly set for each group using the <a href="https://seaborn.pydata.org/tutorial/colour_palettes.html">seaborn husl palette</a>. In the config file you can change to a more intense hsl palette or change the desaturation parameter.</p>
<h3 id="specifying-figure-width">Specifying figure width</h3>
<p>Lovis4u tries to set an optimal figure width taking into account the nucleotide size of visualisation window. You can change it in two ways:<br />
1) Using the <code>--mm-per-nt &lt;float value&gt;</code> argument changing scale which defines given space for each nt cell on canvas. Default: 0.0022 - 0.02, depending on window size.<br />
2) With <code>-fw, --figure-width &lt;float value [cm]&gt;</code> parameter which defines the total output figure width in cm.  </p>
<p>Here we demonstrate usage by plotting a compact visualisation of full loci together with <code>--show-first-feature-label-for</code> argument with an empty list, so not labelling the first occurrence of conserved genes. Note that most of the coloured, variable genes are not labeled in this figure as they are annotated as ‘hypothetical protein’, which LoVis4u ignores to make figures more compact and informative.</p>
<p><div class="language-sh highlight"><pre><span></span><code><span id="__span-7-1"><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>lovis4u<span class="w"> </span>-gff<span class="w"> </span>lovis4u_data/guide/gff_files<span class="w"> </span>-hl<span class="w"> </span>-o<span class="w"> </span>width_test<span class="w"> </span>--show-first-feature-label-for<span class="w"> </span>--figure-width<span class="w"> </span><span class="m">7</span>
</span></code></pre></div>
<a class="glightbox" href="img/lovis4u_set_width.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="f8" loading="lazy" src="img/lovis4u_set_width.png" width="100%" /></a>  </p>

              
            </article>
            
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../../Pipeline/pipeline/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Pipeline description" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Pipeline description
            </div>
          </div>
        </a>
      
      
        
        <a href="../../Parameters/cmd_parameters/" class="md-footer__link md-footer__link--next" aria-label="Next: Command-line parameters" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Command-line parameters
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../..", "features": ["content.code.annotate", "announce.dismiss", "navigation.tabs", "toc.integrate", "header.autohide"], "search": "../../assets/javascripts/workers/search.b97dbffb.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.c99f48ec.min.js"></script>
      
        <script src="../../assets/_markdown_exec_pyodide.js"></script>
      
        <script src="../../javascripts/katex.js"></script>
      
        <script src="https://unpkg.com/katex@0/dist/katex.min.js"></script>
      
        <script src="https://unpkg.com/katex@0/dist/contrib/auto-render.min.js"></script>
      
    
  <script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(() => { lightbox.reload() });
</script></body>
</html>